### DISCUSSION TITLE
Proper use of MallocMessageBuilder(ArrayPtr<>) when the data is meant to persist

### DISCUSSION THREAD

**Discussion by ericjacobs** [capnproto/capnproto#2087] [Q&A] [answered]
Hi,
I'm trying to use MallocMessageBuilder to build a message directly into SHM - so it has to persist beyond the lifetime of the builder, but I'm running into an issue since it goes ahead and zeroes out everything in the destructor.

I'm not understanding why it zeroes out in the destructor rather than the constructor (or likely better; just leaving it up to the user to handling zeroing.) Is there a better way to do this?

(Except for that one aspect, the semantics of MallocMesssageBuilder are exactly what I need: 99% of my messages will fit in the single preallocated segment directly in SHM, and those few that have to spillover into the heap can be dealt with as an exceptional case.)

**Comment by kentonv** [comment, marked as answer]
MallocMessageBuilder allows you to pass one pre-allocated array to the constructor, to use for the first segment, in which case it will only actually malloc after that first segment fills up. It requires that this array is zero'd in advance. The destructor zeros the array because it assumes you may want to reuse the same array again for another message in the future. This is meant as a convenience.

If your goal is to construct a message directly in some memory location, MallocMessageBuilder is not the right tool for this. MallocMessageBuilder's goal -- as its name suggests -- is to allocate memory using malloc(). The ability to pass in pre-allocated scratch space for the first segment is meant as a convenient optimization.

What you should do instead is define your own class, subclassing `MesasgeBuilder`. This shouldn't be too hard. There's only one method you need to implement:

```
virtual kj::ArrayPtr<word> allocateSegment(uint minimumSize) = 0;
```

**Comment by ericjacobs** [comment]
Got it, thank you very much.
